### PR TITLE
Refactor themes/fix night mode

### LINKED ISSUE #10619: [Bug] Theme sometimes do not follow system dark mode on custom or auto schedule
###### Reproduction Steps

1. Use dark mode on custom or auto schedule
2. Disable/Enable dark mode manually

https://user-images.githubusercontent.com/69634269/160028344-bd1ac80b-c9dd-4a93-a8cc-b993d3fd1d16.mp4

###### Expected Result

Theme follow system correctly

###### Actual Result

Theme do not follow system

###### Debug info
```
AnkiDroid Version = 2.16alpha53

Android Version = 11

Manufacturer = samsung

Model = SM-N770F

Hardware = exynos9810

Webview User Agent = Mozilla/5.0 (Linux; Android 11; SM-N770F Build/RP1A.200720.012; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/99.0.4844.73 Mobile Safari/537.36

ACRA UUID = 037bbc86-38a4-4ad7-9c42-05870e9b6086

Scheduler = std2

Crash Reports Enabled = true

DatabaseV2 Enabled = true
```

###### Research
*Enter an [x] character to confirm the points below:*

- [X] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] I have searched for similar existing issues here and on the user forum
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description
Most of Themes.kt code is mine and I find it ugly. Since I'm a wiser man now, I'll try to improve it.

## Fixes
Fixes #10619

## Approach
On the commits (kinda hard to split them, but let me know if anything isn't clear)

## How Has This Been Tested?

On my phone (API 31)

https://user-images.githubusercontent.com/69634269/172143330-90de628c-a12b-45a6-9ee5-0d2a356e0b20.mp4

Upgrading from 2.15.6. Good news is themes aren't becoming corrupted, although I don't know why the previous theme (plain) didn't persist

https://user-images.githubusercontent.com/69634269/172143421-7c7afd8a-5292-4f4e-927f-2b51d14961e1.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt [hunk 1]
```diff
@@ -148,6 +148,29 @@ object PreferenceUpgradeService {
                     preferences.edit().remove("useBackup").apply()
                     preferences.edit().remove("intentAdditionInstantAdd").apply()
                 }
+                // Update Day and Night theme values to their equivalent App theme ones
+                // at @array/app_theme_values on res/values/constants.xml
+                if (previousVersionCode < 21600168) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:153]: This should be in `PreferenceUpgradeService`. 

Version codes no longer work due to our split between architectures.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 2]
```diff
@@ -895,30 +897,43 @@ class Preferences : AnkiActivity() {
                 appThemePref.entryValues = appThemesValues.sliceArray(1..appThemesValues.lastIndex)
             }
 
-            val followSystem = Themes.themeFollowsSystem(requireContext())
-            dayThemePref.isEnabled = followSystem
-            nightThemePref.isEnabled = followSystem
+            (appThemePref.value == Themes.FOLLOW_SYSTEM_MODE).also {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:900]: style: I find this harder to read.

File: AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt [hunk 3]
```diff
@@ -895,30 +897,43 @@ class Preferences : AnkiActivity() {
...
+                dayThemePref.isEnabled = it
+                nightThemePref.isEnabled = it
+            }
 
             appThemePref.setOnPreferenceChangeListener { _, newValue ->
-                dayThemePref.isEnabled = newValue == Themes.FOLLOW_SYSTEM_MODE
-                nightThemePref.isEnabled = newValue == Themes.FOLLOW_SYSTEM_MODE
+                (newValue == Themes.FOLLOW_SYSTEM_MODE).also {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/Preferences.kt:906]: ditto

**Review by david-allison — CHANGES_REQUESTED** [review]
One change needed

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt (hunk 1 above)
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:153]: This should be in a new 'upgrade' class. 

`versionCode` no longer works due to changing how our version numbering works

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:153]: That should probably be added in a comment here in the code so future developers are not lured in, and/or could use a blurb in developer guide "how to do a preferences upgrade"...

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:153]: And I should have read the whole file. Thanks.

**Comment by BrayanDSO** [comment]
Upgrade works now. Thanks.

https://user-images.githubusercontent.com/69634269/172232778-8736b971-5444-4ac0-b40b-f7e70991e7ac.mp4

**Comment by BrayanDSO** [comment]
Drafting while I fix the test

**Comment by BrayanDSO** [comment]
A thought, if we go _downgrade_ the app, it will corrupt the theme, since the new theme values can't be used on the older code, which requires cleaning the app's data to make it work again OR go back to the new version.

How much of a problem is this? (i.e. should I handle downgrades or not?)

**Comment by BrayanDSO** [comment]
Deconflicted and fixed the problem with systemIsInNightMode initialization on tests with `@EmptyApplication`

File: AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.kt [hunk 4]
```diff
@@ -38,22 +38,22 @@ class CardAppearance(private val customFonts: ReviewerCustomFonts, private val c
...
             cardClass.append(" vertically_centered")
         }
-        if (isNightMode) {
+        if (currentTheme.isDark) {
             // Enable the night-mode class
             cardClass.append(" night_mode nightMode")
 
             // Emit the dark_mode selector to allow dark theme overrides
-            if (currentTheme == Themes.APP_DARK_THEME) {
+            if (currentTheme == Theme.DARK) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.kt:51]: This is confusing. We have `isDark` and `== DARK` meaning two different things.

Maybe `isNightMode`?

Note: we do want to remove this code fairly soon

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt [hunk 5]
```diff
@@ -279,5 +281,27 @@ object PreferenceUpgradeService {
...
+                    }
+                    if (nightTheme == "1") { // dark
+                        putString("nightTheme", "4")
+                    } else { // black
+                        putString("nightTheme", "3")
+                    }
+                }
+                if (AnkiDroidApp.isInitialized()) {
+                    Themes.updateCurrentTheme()
+                }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:303]: can theoretically remove `invertedColors` as well?

**Review by david-allison — COMMENTED** [review]
Two questions, but generally looks good

File: AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.kt (hunk 4 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/cardviewer/CardAppearance.kt:51]: Have thought about it and  `current theme is night mode` didn't sound  very natural to me. But it's not something I care much, so I'll do the change.

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt (hunk 5 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:303]: Sure

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt [hunk 6]
```diff
@@ -279,5 +281,28 @@ object PreferenceUpgradeService {
...
             }
         }
+        internal class UpgradeDayAndNightThemes : PreferenceUpgrade(5) {
+            override fun upgrade(preferences: SharedPreferences) {
+                val dayTheme = preferences.getString("dayTheme", "0")
+                val nightTheme = preferences.getString("nightTheme", "0")
+
+                preferences.edit {
+                    if (dayTheme == "1") { // plain
+                        putString("dayTheme", "2")
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:291]: Might this be a good time to use more descriptive values than stuff like "2"?

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:291]: IMO the values don't matter. 

And changing them would mean having to upgrade the app theme values as well, which is an unnecessary work.

**Review by david-allison — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt (hunk 6 above)
Comment (oakkitten) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/PreferenceUpgradeService.kt:291]: It's not that this really matters, but [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) are usually regarded as anti-pattern.

**Comment by BrayanDSO** [comment]
Rebased to solve conflicts

**Review by mikehardy — APPROVED** [review]
LGTM - just ran it through CI tasks locally after rebase to make sure, all good

**Pull request merged**